[WIP] Start on adding C type sizes

jyn514 opened the pull request:
Closes #47 

- ~~Rename `PointerWidth` to `Size` and use it for other types as well~~
- Add new `Size` enum
- Add `CDataModel` and `data_model()` function to `impl Triple`

My main question is how to tell what data model belongs to which target. I have some very broad heuristics but there's too many possible triples to match them all exhaustively (over 200000!)

sunfishcode commented on src/triple.rs:
```diff
@@ -17,41 +17,40 @@ pub enum Endianness {
     U16,
     U32,
     U64,
 }
```
I'd like to propose we introduce a new type rather than repurposing `PointerWidth`. `PointerWidth` is one of the few non-`non_exhaustive` enums in target-lexicon, and it's convenient for the very common case of just asking the library "is this platform 32-bit or 64-bit (or 16-bit)?" since that bit of information is needed in a lot of places.

But for C type layout, as we look forward to adding support for struct layout, but also SIMD types, minifloats, `int128`, _Complex, long double, _Complex long double, and oh, the places we'll go ;-), we'll want something to extend to new uses. So I suggest a new `Size` struct with a private integer field here, which can hold any size.

jyn514 replied:
I'm happy to make a new enum, I felt a little uncomfortable removing `non_exhaustive` anyway.

> I suggest a new Size struct with a private integer field here, which can hold any size.

Ok, if we made it a struct I could also add `char_bit` which would make `bytes()` more useful. I think the `bytes()` and `bits()` API is still useful, I'd like to keep it.

jyn514 commented on src/triple.rs:
```diff
@@ -70,6 +69,70 @@ pub enum CallingConvention {
+    pub fn pointer_width(&self) -> Size {
+        match self {
+            CDataModel::LP32 | CDataModel::ILP32 => Size::U32,
+            CDataModel::LLP64 | CDataModel::LP64 | CDataModel::ILP64 => Size::U64,
```
`non_exhaustive` only impacts downstream users, not the library it's defined in. Personally this is expected behavior for me, this ensures that if we add a new `CDataModel` variant we'll have to add it to `pointer_width()`, etc. before publishing. This is the same reason I used a match in `long_long_size` instead of a plain `CDataModel::U64`.

sunfishcode replied:
Ah, that make sense!

jyn514 commented:
Addressed review comments. Still a little unsure what data models correspond to which targets, but I guess more of those can be added in a follow-up PR.

sunfishcode commented on src/data_model.rs:
```diff
@@ -0,0 +1,105 @@
+    /// Return the number of bytes in a size.
+    ///
+    /// Depending on the target, a byte may not necessarily be 8 bits.
+    pub fn bytes(self) -> u8 {
```
There are so many places that assume "byte" is "8-bit byte" these days that I'm worried this distinction is too subtle. What would you think about calling this method `c_chars` instead of `bytes` and said "Return the size in units of C `char`s (as C's `sizeof` does)"?

sunfishcode commented on src/data_model.rs:
```diff
@@ -0,0 +1,105 @@
+        bytes
+    }
+    /// Create a new `Size`, assuming that `char_bit` is 8.
+    fn with_bits(bits: u8) -> Self {
```
If we're going to support `CHAR_BIT!=8`, it seems like we should consistently avoid assuming `CHAR_BIT=8`. How inconvenient would it be to add a `char_bit` parameter here?

jyn514 replied:
My idea was that this would be an internal API and that CHAR_BIT would default to 8. You can still construct `Size` with a different bit width manually.

However, after looking at your comments below I'm starting to wonder the utility of having a configurable CHAR_BIT. If we can't even name `bytes()` after the number of bits in a byte, is it worth having the distinction? Do you know any targets rustc supports with CHAR_BIT != 8?

sunfishcode replied:
No, I don't know of any targets in rustc with `CHAR_BIT != 8`. I think it is reasonable to have the API assume `CHAR_BIT == 8`, but I'd like to either make that assumption everywhere, or not at all. If we assume `CHAR_BIT` is 8, then `Size` shouldn't have a `char_bit` field. If we don't assume it, then we shouldn't have any implicit 8-bit assumptions.

jyn514 replied:
I'm ok with getting rid of char_bit, but in that case we should either change Size to `enum { 8, 16, 32, 64 }` or return an `Option` from bytes. If there's a size that's not a multiple of 8, there's nothing to do right now except panic.

jyn514 replied:
Actually I guess there is a 3rd option which is to round up to the nearest multiple of 8.

sunfishcode replied:
A panic sounds reasonable to me here. An enum would also be fine.

jyn514 replied:
Ok, I changed it to an enum.

sunfishcode commented on src/data_model.rs:
```diff
@@ -0,0 +1,105 @@
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum CDataModel {
+    /// The data model used most commonly on Win16. `int` and `long` are 32 bits.
```
Is this comment correct? I'm not familiar with `LP32` specifically, but if `int` is 32 bits, why isn't this `ILP32`?

jyn514 replied:
This comment is incorrect, `int` is 16 bit on LP32.

sunfishcode commented on src/triple.rs:
```diff
@@ -127,6 +127,35 @@ impl Triple {
+            } else {
+                Err(())
+            }
+        } else {
```
This `else` is for `PointerWIdth::U16`? Could you switch the if-else chain here to a `match` so that it's more explicit?

Also, I'm confused about why a 16-bit pointer system uses a data model with `P32` in it.

jyn514 replied:
I'm not sure why it uses 32-bit pointers either, but all the sources I've found seem to back that up. I haven't found anything official, though.

I'll see if I can make the code a little more readable with a `match`.

- https://en.cppreference.com/w/cpp/language/types
- http://nickdesaulniers.github.io/blog/2016/05/30/data-models-and-word-size/
- http://dewkumar.blogspot.com/2012/03/cc-data-model-3264-bit.html

bjorn3 commented on src/triple.rs:
```diff
@@ -127,6 +127,38 @@ impl Triple {
+            }
+            PointerWidth::U16 => {
+                if self.operating_system == OperatingSystem::Windows {
+                    Ok(CDataModel::LP32)
```
According to https://en.m.wikipedia.org/wiki/16-bit_computing there is a difference between near (16bit) and far (32bit) pointers on 16bit x86. There are multiple code models that use a different mix of near and far pointers for code and data.

jyn514 replied:
This sounds like a little more than I want to tackle in a first PR, especially since target_lexicon doesn't currently distinguish between code and data pointers. Are you good with me removing this U16 branch altogether and just returning `Err`?

sunfishcode replied:
Yes, that sounds good.

jyn514 commented:
How does this look?
